File: src/priorityq.py
"""Create Priority Q object."""


class PriorityQueue(object):
    """Create a new PriorityQueue object."""

    def __init__(self):
        """Initialized value of a Priority Queue."""
        self._priority_dict = {}

    def insert(self, value, priority=0):
        """Insert a new value with priority zero unless otherwise stated."""
        self._priority_dict.setdefault(priority, []).append(value)

    def pop(self):
        """Pop first highest priority value."""
        highest_priority = max(self._priority_dict, key=int)
        return self._priority_dict[highest_priority].pop(0)

    def peek(self):
        """Show the first highest priority value."""
        highest_priority = max(self._priority_dict, key=int)
        return self._priority_dict[highest_priority][0]

File: js/priorityq.js
class PriorityQueue {
    constructor() {
        this.priority_dict = {};
        this.max = null;
    }

    insert(data, priority=0) {
        if(this.priority_dict[priority] === undefined) {
            this.priority_dict[priority] = [data];
        } else {
            this.priority_dict[priority].push(data);
        }
        this.getMax();
    }

    getMax() {
        let priority_list = Object.keys(this.priority_dict);
        if(priority_list.length != 0) {
            let max = 0;
            for(let i = 0; i < priority_list.length; i++) {
                if (priority_list[i] > max) this.max = parseInt(priority_list[i]);
            }
        } else {
            this.max = null;
        }
    }

    pop() {
        if(this.max === null) {
            return 'Nothing to pop';
        } else {
            let popped = this.priority_dict[this.max].shift();
            if(this.priority_dict[this.max].length === 0) {
                delete this.priority_dict[this.max];
                this.getMax();
            }
            return popped;
        }
    }

    peek() {
        if(this.max === null) {
            return 'None';
        } else {
            return this.priority_dict[this.max][0];
        }
    }
}

module.exports = PriorityQueue;